src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+
+        if (name.find(this->modelName) != std::string::npos)
+        {
+            this->filteredCollisionEntities.push_back(entity);
+        }
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::InterpolateData(
+                    const ignition::msgs::Contact &contact)
+{
+    if (contact.position_size() == 4)
+    {
+        ignition::math::Vector3d contact1 =
+            ignition::msgs::Convert(contact.position(0));
+        ignition::math::Vector3d contact2 =
+            ignition::msgs::Convert(contact.position(1));
+        ignition::math::Vector3d contact3 =
+            ignition::msgs::Convert(contact.position(2));
+        ignition::math::Vector3d contact4 =
+            ignition::msgs::Convert(contact.position(3));
+
+        ignition::math::Vector3d direction1 =
+            (contact2 - contact1).Normalized() * this->resolution;
+        ignition::math::Vector3d direction2 =
+            (contact4 - contact1).Normalized() * this->resolution;
+
+        ignition::math::Vector3d interpolatedVector = contact1;
+        this->interpolatedPosition.push_back(contact1);
+
+        // auxiliary Vector3d to iterate through contacts
+        ignition::math::Vector3d tempVector(0.0, 0.0, 0.0);
+
+        uint64_t steps1 =
+            contact1.Distance(contact2) / this->resolution;
+        uint64_t steps2 =
+            contact1.Distance(contact4) / this->resolution;
+
+        for (uint64_t index1 = 0; index1 <= steps1; ++index1)
+        {
+            tempVector = interpolatedVector;
+            for (uint64_t index2 = 0; index2 < steps2; ++index2)
+            {
+                interpolatedVector += direction2;
+                this->interpolatedPosition.push_back(interpolatedVector);
+            }
+            if (index1 != steps1)
+            {
+                interpolatedVector = tempVector;
+                interpolatedVector += direction1;
+                this->interpolatedPosition.push_back(interpolatedVector);
+            }
+        }
+    }
+}
+
+//////////////////////////////////////////////////
+OpticalTactilePlugin::OpticalTactilePlugin()
+            : System(), dataPtr(std::make_unique<OpticalTactilePluginPrivate>())
+{
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::Configure(const Entity &_entity,
+                            const std::shared_ptr<const sdf::Element> &_sdf,
+                            EntityComponentManager &_ecm,
+                            EventManager &)
+{
+    this->dataPtr->sdfConfig = _sdf->Clone();
+    this->dataPtr->model = Model(_entity);
+
+    if (!this->dataPtr->model.Valid(_ecm))
+    {
+      ignerr << "Touch plugin should be attached to a model entity. "
+            << "Failed to initialize." << std::endl;
+      return;
+    }
+
+    // Configure subscriber for Depth Camera images
+    sdf::ElementPtr depthCameraSdf =
+      _sdf->GetParent()->GetElement("link")->GetElement("sensor");
```

Comment:
This seems like it's looking for the `<sensor>` tag in the first `<link>` element. What if the sensor is in a different link? You probably want to either specify the sensor name in the plugin or look for entities that have either `components::DepthCamera` or `components::RgbdCamera`.

Reply:
So the plugin is supposed to have one link only, with one depth camera and one contact sensor. Otherwise, it wouldn't make sense. I should still be able to differentiate between those two kind of sensors, because they could be placed in different order. Right now it takes the first sensor it finds.

Apart from that, maybe I'm missing checks on how the `sdf` model is built (one link with two specific sensors)?

Reply:
I see. I didn't know about the one link assumption. I was thinking that a user can add this system to any model that has a depth camera sensor. I'm good with keeping it simple for now and adding more configuration options later. I would suggest adding a TODO about it though.